### PR TITLE
LevelDb fully pruned block store

### PR DESCRIPTION
**Pull request by n1bor**
and change to test logging to allow
unit tests to work on Travis. Without the log
change the unit tests fail due to hitting 4 Meg limit

### CODE REVIEW

**Comment by mikehearn** [comment]
Why do you keep closing and reopening the pull request? Is there some issue with github? You can force push over the branch you are pull requesting to update it.

So - very initial comments without reading the code in much detail:

* Can you give performance numbers vs H2? How much of an improvement is it?
* I see that we're using a Java reimplementation of LevelDB rather than the original. What's the reason for doing that vs using the JNI bindings?
* Watch out for code style. Please always try and match the code style of projects you are contributing to - in this case, if you look at the diff you'll see that indentation is more or less a random mishmash. Probably you are using Eclipse. Configure it to use spaces not tabs and then reformat to ensure it matches. Likewise, spaces around = signs and other things.
* I saw you're adding your own LRU cache and specifying your own Bloom filters. Is that really needed? LevelDB C++ already does its own caching and it shouldn't be needed to specify custom Bloom filters. At least I didn't need to do this when I implemented LevelDB for Bitcoin Core.
* Can you split out the Travis change to a separate pull request please, so it can be reviewed and merged separately?

Thanks!

File: core/pom.xml [hunk 1]
```diff
@@ -470,6 +477,11 @@
             <artifactId>orchid</artifactId>
             <version>1.1</version>
         </dependency>
+        <dependency>
```
Comment (schildbach) [inline core/pom.xml]: For each dependency, we need the hash of the artifact in the maven-enforcer-plugin configuration.

**Comment by n1bor** [comment]
In response...
1) H2 sync to block 338000 running now. But think from memory is days not hours!
2) Went with pure java out of ease of integration. Swapping this for JNI should be simple as tricky bit was the transaction/rollback handling. Although JNI might be different!
3) OK - can tidy up.
4) Will rerun benchmark without caches - last test was 5h45 with no cache, 3h45 with bloom and 3h02 with bloom and utxocache (1m) . But fixed many bugs since last test. Bloom was needed as leveldb caching does not help "hasunspentoutputs" as the data is never there (it is check for dup tx hashes). These times were to an SSD - magnetics about 25% slower, on quad core with 12Gig RAM.
5) OK will split travis (did mean too but lost it somehow!) used to clearcase not git!

Below is how many mins it spends in each call with both caches - total time was 182 mins.

image

**Comment by n1bor** [comment]
H2 not good at initial download. Keeps stalling every few hours. But is about 30 times slower than LevelDB once you get to the larger blocks.
H2 Block 180,305 - 186,000 - 70mins
LevelDb Block 180,000 - 186,000- 2mins 30secs. (only takes 20mins from zero to 200,000!).

**Comment by n1bor** [comment]
Below are stats with/without caches:
image

Can not compare to H2 as takes days - is at least 30 times slower... and I have to keep restarting.

Do you want a push with or without the caches? Will then tidy up style, split into 2 commits one with tavis change one with the leveldb change.

Once this merged will look at a getAddressBalance and a getAllUTXOHash just to see fast they are on big data sets.

**Comment by mikehearn** [comment]
OK, I think you proved your point. I'm still curious as to why this happens - but maybe the Java LevelDB reimplementation doesn't do caching or doesn't do it in the same way as the C++ version does.

Anyway it's fine to keep the caches. But please do include some of these data points in comments, so future readers can understand why they are there.

**Comment by mikehearn** [comment]
Hey n1bor, do you plan to revisit this pull req w.r.t the feedback?

**Comment by n1bor** [comment]
Yes - just little tied up.. Did test with native LevelDB and was 30% faster still. Hope to get round to it in week or 2.

**Comment by schildbach** [comment]
@n1bor Are you still interested in this PR?

**Comment by n1bor** [comment]
Seems shame to lose it - let me see if I can tidy up today/tomorrow.

File: core/pom.xml [hunk 2]
```diff
@@ -301,6 +301,13 @@
                 <version>2.12.4</version>
                 <configuration>
                     <argLine>${surefireArgLine}</argLine> <!-- This is required for code coverage to work. -->
+                    <!-- This is added to reduce log output below 4Meg limit of Travis-CI -->
```
Comment (schildbach) [inline core/pom.xml]: I think this should go into a separate PR.

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java [hunk 3]
```diff
@@ -0,0 +1,1070 @@
+package org.bitcoinj.store;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: All the new files are missing the license header.

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java [hunk 4]
```diff
@@ -0,0 +1,1070 @@
...
+ * implementation as the backing data store.<p>
+ * 
+ * <p>Includes number of caches to optimise the initial blockchain download</p> 
+ *  
+ *
+ */
+
+
+public class LevelDbFullPrunedBlockStore implements FullPrunedBlockStore {
+	private static final Logger log = LoggerFactory.getLogger(LevelDbFullPrunedBlockStore.class);
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: Leading tab/space mismatches all over the place. We're using 4 spaces per indent level.

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java [hunk 5]
```diff
@@ -0,0 +1,1070 @@
...
+	
+	//Standard blockstore properties
+    protected Sha256Hash chainHeadHash;
+    protected StoredBlock chainHeadBlock;
+    protected Sha256Hash verifiedChainHeadHash;
+    protected StoredBlock verifiedChainHeadBlock;
+    protected int fullStoreDepth;
+    //Indicates if we track and report runtime for each method
+    //this is very useful to focuse performance tuning on correct areas.
+    protected boolean instrument=false;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: Space before and after operators like this.

**Comment by schildbach** [comment]
I added a couple of comments, mostly code style. I assume Mike was already happy with what the PR does, so I'll merge as soon as the comments are addressed.

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java [hunk 6]
```diff
@@ -0,0 +1,1070 @@
...
+ * <p>Includes number of caches to optimise the initial blockchain download</p> 
+ *  
+ *
+ */
+
+
+public class LevelDbFullPrunedBlockStore implements FullPrunedBlockStore {
+	private static final Logger log = LoggerFactory.getLogger(LevelDbFullPrunedBlockStore.class);
+	
+	NetworkParameters params;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: Should this be `private final`?

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java [hunk 7]
```diff
@@ -0,0 +1,1070 @@
...
+
+	//LevelDB reference.
+	DB db=null;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: private?

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java [hunk 8]
```diff
@@ -0,0 +1,1070 @@
...
+    protected Sha256Hash chainHeadHash;
+    protected StoredBlock chainHeadBlock;
+    protected Sha256Hash verifiedChainHeadHash;
+    protected StoredBlock verifiedChainHeadBlock;
+    protected int fullStoreDepth;
+    //Indicates if we track and report runtime for each method
+    //this is very useful to focuse performance tuning on correct areas.
+    protected boolean instrument=false;
+    //instrumentation stats
+    long wallTimeStart;
```
Comment (schildbach) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: Perhaps use Guava [Stopwatch](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Stopwatch.html) for this?

File: core/pom.xml (hunk 2 above)
Comment (n1bor) [inline core/pom.xml]: removed - will find out if still needed once I push to travis and will add extra PR if needed for just this.

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java (hunk 3 above)
Comment (n1bor) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: added

File: core/pom.xml (hunk 2 above)
Comment (n1bor) [inline core/pom.xml]: Travis working without this so will skip it! Guess the logging has reduced or travis is allowing more log lines?

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java (hunk 4 above)
Comment (n1bor) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: Fixed

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java (hunk 5 above)
Comment (n1bor) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: Fixed

File: core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java (hunk 8 above)
Comment (n1bor) [inline core/src/main/java/org/bitcoinj/store/LevelDbFullPrunedBlockStore.java]: Changed.. Could change the LRU cache too - but will do later..

**Comment by n1bor** [comment]
OK - tidy now. Fixed most comments - missed a couple. Will fix in next release. Currently syncs at about 3m/bytes per sec. So taking 4-5hours.

**Comment by schildbach** [comment]
Finally, I merged this. Many thanks for bearing with us for so long!
